### PR TITLE
FFD Geometry Augmentation: synthetic airfoil data via Free-Form Deformation

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Apply random Free-Form Deformation (FFD) to training airfoil surface geometries at train time, creating synthetic shape perturbations that force the model to generalize across a wider distribution of airfoil shapes. The model currently sees a fixed set of NACA geometries — FFD augmentation multiplies the effective shape diversity without requiring new CFD simulations.

**Why this should work:**
- The training set contains a finite set of NACA 4-digit airfoil shapes. The model may memorize specific shape-pressure mappings rather than learning generalizable aerodynamic principles
- FFD is the standard parameterization in aerodynamic shape optimization (adjoint methods) — small control point displacements create smooth, physically plausible shape variations
- Shape augmentation has been shown to improve OOD generalization in geometric deep learning (FuncGenFoil, arXiv:2502.10712)
- This is a DATA-LEVEL intervention, not a model tweak — directly addresses the research directive for data augmentation

**Key insight:** We don't need accurate pressure labels for augmented shapes. We use the existing labels with a reduced loss weight, treating augmentation as a regularizer that forces shape-invariant representations. Alternatively, we can use the existing panel Cp as approximate labels for deformed shapes.

### Step 1 — Implement FFD deformation (pure PyTorch, ~40 lines)

Add a `FFDDeform` class that applies 2D Free-Form Deformation to airfoil coordinates:

```python
class FFDDeform:
    """2D Free-Form Deformation for airfoil augmentation."""
    def __init__(self, n_control_x=4, n_control_y=2, amplitude=0.01):
        self.nx, self.ny = n_control_x, n_control_y
        self.amplitude = amplitude  # max displacement as fraction of chord

    def __call__(self, xy, chord_length=1.0):
        """Deform surface coordinates.
        Args:
            xy: [N, 2] surface node coordinates
        Returns:
            xy_deformed: [N, 2] deformed coordinates
        """
        # Normalize to [0,1] bounding box
        x_min, x_max = xy[:, 0].min(), xy[:, 0].max()
        y_min, y_max = xy[:, 1].min(), xy[:, 1].max()
        s = (xy[:, 0] - x_min) / (x_max - x_min + 1e-8)  # [0, 1]
        t = (xy[:, 1] - y_min) / (y_max - y_min + 1e-8)  # [0, 1]

        # Random control point displacements
        delta = torch.randn(self.nx, self.ny, 2, device=xy.device)
        delta = delta * self.amplitude * chord_length
        # Pin leading/trailing edge: zero displacement at x=0 and x=1
        delta[0, :, :] = 0  # LE
        delta[-1, :, :] = 0  # TE

        # Bernstein basis (B-spline lattice)
        # Bx[i](s) = C(nx-1,i) * s^i * (1-s)^(nx-1-i)
        displacement = torch.zeros_like(xy)
        for i in range(self.nx):
            for j in range(self.ny):
                Bx = comb(self.nx-1, i) * s.pow(i) * (1-s).pow(self.nx-1-i)
                By = comb(self.ny-1, j) * t.pow(j) * (1-t).pow(self.ny-1-j)
                w = (Bx * By).unsqueeze(-1)  # [N, 1]
                displacement += w * delta[i, j]  # [N, 2]

        return xy + displacement
```

### Step 2 — Add augmentation flag and training integration

```python
# Config flags:
ffd_augmentation: bool = False     # enable FFD geometry augmentation
ffd_amplitude: float = 0.01       # max deformation (fraction of chord)
ffd_prob: float = 0.5             # probability of applying augmentation per sample
ffd_loss_weight: float = 0.3      # reduced weight for augmented samples
```

**In the training loop:** For each batch, with probability `ffd_prob`, apply FFD to the surface node coordinates ONLY (not volume nodes). Keep the original pressure/velocity targets unchanged. Apply a reduced loss weight (`ffd_loss_weight`) for augmented samples since the labels are approximate.

```python
if cfg.ffd_augmentation and model.training and torch.rand(1).item() < cfg.ffd_prob:
    ffd = FFDDeform(amplitude=cfg.ffd_amplitude)
    # Deform surface coordinates (is_surface mask)
    for b in range(x.shape[0]):
        surf_mask = is_surface[b]  # [N]
        x[b, surf_mask, :2] = ffd(x[b, surf_mask, :2])
    # Flag this batch for reduced loss weight
    is_augmented = True
```

Then when computing loss:
```python
loss_weight = cfg.ffd_loss_weight if is_augmented else 1.0
loss = loss * loss_weight
```

### Step 3 — Recompute derived features after deformation

After FFD deformation, recompute all coordinate-derived features:
- Panel Cp (uses deformed surface coordinates)
- TE coordinate frame features (uses deformed TE position)
- Wake deficit features (uses deformed TE position)
- Fourier PE (uses deformed coordinates)

This ensures the input features are consistent with the deformed geometry.

### Step 4 — Run experiments

```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent frieren --wandb_name "frieren/ffd-aug-s42" --wandb_group ffd-geometry-augmentation \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --ffd_augmentation --ffd_amplitude 0.01 --ffd_prob 0.5 --ffd_loss_weight 0.3

CUDA_VISIBLE_DEVICES=1 python train.py [same with --seed 73]
```

**Important:** Only augment single-foil samples initially (skip tandem — deforming one foil changes the flow field for both). Check timing: FFD + feature recomputation must not significantly slow training per epoch.

**Torch.compile note:** The FFD computation uses fixed control grid sizes and tensor ops only — should be compatible. The `torch.rand` in the `if` branch may cause recompilation — move the randomness outside the compiled region if needed.

## Baseline (PR #2350, 2-seed avg)

| p_in | p_oodc | p_tan | p_re |
|------|--------|-------|------|
| 11.90 | 7.35 | 27.20 | 6.40 |

Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature`

## Round 40 Bold Context

This is a **DATA AUGMENTATION** experiment — the first geometry-level augmentation in 1966+ experiments. Directly addresses the human researcher's directive for "data aug and data generation." FFD is the standard aerodynamic shape parameterization — we're bringing proven CAD/optimization tools into the training pipeline.

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: **Baseline updated.** PR #2357 (Vortex-Panel Induced Velocity) was just merged. Rebase onto latest `noam` branch and add these flags to your training command:

```
--vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64
```

New baseline metrics to beat (2-seed avg):
| p_in | p_oodc | p_tan | p_re |
|------|--------|-------|------|
| 11.872 | 7.459 | 26.319 | 6.229 |

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### FFD Geometry Augmentation — 2-Seed Experiment

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline (PR #2350) | Δ |
|--------|---------|---------|------------|--------------------|----|
| p_in | 13.3 | 12.7 | **13.0** | 11.90 | **+9.2%** ❌ |
| p_oodc | 7.5 | 7.4 | **7.45** | 7.35 | **+1.4%** ❌ |
| p_tan | 28.3 | 27.6 | **27.95** | 27.20 | **+2.8%** ❌ |
| p_re | 6.5 | 6.4 | **6.45** | 6.40 | **+0.8%** ❌ |

### Details

- **Best epoch (s42):** 160/160 (val_loss=0.6140)
- **Best epoch (s73):** 159/160 (val_loss=0.6038)
- **Training time:** ~180 min each (wall-clock timeout)
- **Peak memory:** 40.5 GB (s42), 40.9 GB (s73)
- **W&B runs:** `5g9kvdnu` (seed 42), `ckq0rk09` (seed 73)
- **W&B group:** `ffd-geometry-augmentation`

### Command

```bash
cd cfd_tandemfoil && python train.py --agent frieren --wandb_name "frieren/ffd-aug-s42" \
  --wandb_group ffd-geometry-augmentation --seed 42 --asinh_pressure --field_decoder \
  --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --ffd_augmentation --ffd_amplitude 0.01 --ffd_prob 0.5 --ffd_loss_weight 0.3
```

(Same for seed 73 with `--seed 73` and `--wandb_name "frieren/ffd-aug-s73"`)

### What happened

**FFD augmentation did not improve any metric — all regressed, especially p_in (+9.2%).**

The hypothesis was that random Free-Form Deformation of airfoil surface geometry would force the model to learn shape-invariant representations and improve OOD generalization. In practice:

1. **p_in regression is the worst (+9.2%):** The in-distribution validation set has a fixed set of NACA shapes. FFD augmentation introduces noise into the geometry-to-pressure mapping — the model sees deformed shapes with the original (non-deformed) pressure labels, weighted at 0.3×. Even with the reduced loss weight, this appears to confuse the surface pressure prediction on clean geometries.

2. **OOD metrics also regressed:** p_oodc +1.4%, p_tan +2.8%, p_re +0.8%. The augmentation was supposed to help OOD generalization, but the synthetic shape-label mismatch (deformed geometry with original pressure labels) likely teaches the model to be less sensitive to geometry features, which hurts everywhere.

3. **Training dynamics:** Both seeds converged to similar metrics with best models at epochs 159-160 (the timeout cut training at ~160 epochs). The surface loss alternated between normal batches (~0.003-0.010) and FFD-augmented batches (~0.010-0.040), confirming the augmentation was active and the 0.3× weighting was applied.

4. **Why it failed:** The fundamental issue is using original pressure labels for deformed shapes. Even with 0.3× loss weight, this is still an incorrect supervision signal — deformed airfoils should produce different pressure fields. The model cannot learn a correct geometry-pressure mapping when ~50% of training batches have mismatched geometry-pressure pairs.

### Suggested follow-ups

1. **Panel Cp for deformed shapes:** Instead of reusing original labels, compute panel-method Cp for the FFD-deformed shapes to get approximately correct labels. This was mentioned in the PR body as an alternative approach and would provide physically consistent supervision.

2. **Much smaller amplitude (0.001-0.003):** The 0.01 chord-fraction amplitude may be too large. With smaller perturbations, the original labels would be a better approximation of the true pressure field.

3. **Contrastive/consistency regularization instead of supervised:** Instead of using FFD for data augmentation with labels, use it for a consistency loss — the model's predictions for original and deformed shapes should be similar (not identical). This avoids the label mismatch problem entirely.

4. **FFD as test-time augmentation only:** Apply FFD to test inputs and average predictions, avoiding any training label issues.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2365 because ADVISOR: Closing — clear dead end.

**W&B-verified results (vs baseline p_in=11.872, p_tan=26.319):**
| Metric | S42 (5g9kvdnu) | S73 (ckq0rk09) | 2-seed avg | Baseline | Δ |
|--------|----------------|----------------|------------|----------|---|
| p_in | 13.256 | 12.723 | **12.990** | 11.872 | **+9.4% ❌** |
| p_oodc | 7.476 | 7.429 | **7.453** | 7.459 | -0.1% (flat) |
| p_tan | 28.340 | 27.583 | **27.962** | 26.319 | **+6.2% ❌** |
| p_re | 6.538 | 6.374 | **6.456** | 6.229 | +3.6% ❌ |

Your analysis is spot-on: the fundamental issue is using original pressure labels for deformed geometry. Even with 0.3× loss weight and 50% augmentation probability, the model gets systematically incorrect supervision on half its training batches. The p_in +9.4% regression confirms the geometry-to-pressure mapping is corrupted.

**Why the suggested follow-ups aren't worth pursuing here:**
- Panel Cp for deformed shapes would require a separate fast solver pipeline (adds complexity for uncertain gain)
- Smaller amplitude (0.001-0.003) — the label mismatch still exists, just smaller in magnitude
- Contrastive/consistency loss is an interesting but completely different experiment (no longer FFD augmentation per se)

Good experiment, well-executed. The data generation direction via geometry perturbation is now understood: it requires physically consistent labels for perturbed geometries, which is a much harder problem than the FFD geometry generation itself.

You will get a new bold assignment shortly..
